builtin/logical/database/rotation.go

```diff
@@ -0,0 +1,380 @@
+package database
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/sdk/queue"
+)
+
+const (
+	// interval to check the queue for items needing rotation
+	queueTickSeconds  = 5
+	queueTickInterval = queueTickSeconds * time.Second
+
+	// wal storage key used for static account rotations
+	staticWALKey = "staticRotationKey"
+)
+
+// populateQueue loads the priority queue with existing static accounts. This
+// occurs at initialization, after any WAL entries of failed or interrupted
+// rotations have been processed. It lists the roles from storage and searches
+// for any that have an associated static account, then adds them to the
+// priority queue for rotations.
+func (b *databaseBackend) populateQueue(ctx context.Context, s logical.Storage) {
+	log := b.Logger()
+	log.Info("populating role rotation queue")
+
+	// build map of role name / wal entries
+	walMap, err := b.loadStaticWALs(ctx, s)
+	if err != nil {
+		log.Warn("unable to load rotation WALs", "error", err)
+	}
+
+	roles, err := s.List(ctx, databaseStaticRolePath)
+	if err != nil {
+		log.Warn("unable to list role for enqueueing", "error", err)
+		return
+	}
+
+	for _, roleName := range roles {
+		select {
+		case <-ctx.Done():
+			log.Info("rotation queue restore cancelled")
+			return
+		default:
+		}
+
+		role, err := b.StaticRole(ctx, s, roleName)
+		if err != nil {
+			log.Warn("unable to read static role", "error", err, "role", roleName)
+			continue
+		}
+		walEntry := walMap[roleName]
+		if role == nil || role.StaticAccount == nil {
+			if walEntry != nil {
+				// no static role exists for this wal entry, delete it
+				if err := framework.DeleteWAL(ctx, s, walEntry.walID); err != nil {
```

Comment:
I think this logic is already happening in the loadStaticWALs function? Could we dedup either here or there?

Reply:
It is, great catch thanks. Removed 94b9106ac